### PR TITLE
CC 2873

### PR DESCRIPTION
**Pull request by ben-abraham**
Redesign/Rewrite of API.
Not completed yet, but almost completely functional.

### CODE REVIEW

File: ConstructorIO.Test/ListItemTests.cs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+using System.Collections.Generic;
+
+namespace ConstructorIO.Test
+{
+    [TestClass]
+    public class ListItemTests
+    {
+        public ConstructorIOAPI MakeAPI()
+        {
+            return new ConstructorIOAPI("UuXqlIafeKvwop6DaRwP", "5tHZR5xflF6bNLgvpa60");
```
Comment (danmcc) [inline ConstructorIO.Test/ListItemTests.cs]: Can we pass these in with environment vars?

Comment (ben-abraham) [inline ConstructorIO.Test/ListItemTests.cs]: Yes, I will do that.

File: constructor.io/ListItemAutocompleteType.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+namespace ConstructorIO
+{
+    public enum ListItemAutocompleteType
+    {
+        [StringValue("Products")]
+        Products,
+        [StringValue("Search Suggestions")]
+        SearchSuggestions
+    }
+}
```
Comment (danmcc) [inline constructor.io/ListItemAutocompleteType.cs]: The autocomplete sections can actually have arbitrary names, so can we just make this a string?  Also, can we call them "sections" instead of "types"?  That's how we should refer to them elsewhere.

Comment (ben-abraham) [inline constructor.io/ListItemAutocompleteType.cs]: The constructors and functions take both ListItemAutocompleteType (which I will rename) as well as string, so both of the following are valid.
`
var a = new ListItem("Test", ListItemAutocompleteType.Products")
var b = new ListItem("Test", "Arbitrary Section")
`

File: constructor.io/ListItem.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+using System.Text;
+
+namespace ConstructorIO
+{
+    using HashArgs = Dictionary<string, object>;
+
+    public class ListItem
+    {
+        private string _name;
+        private string _category;
```
Comment (danmcc) [inline constructor.io/ListItem.cs]: Can we call this "autocomplete section" instead of "category" everywhere, just to be consistent with the API nomenclature?

File: constructor.io/ConstructorIOAPI.cs [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+            if (verifyResponse.Item2.IndexOf("successful authentication") != -1)
+            {
+                //TODO: beter way to do this?
+                return true;
+            }
+            return false;
+        }
+
+        public async Task<bool> AddAsync(ListItem Item, bool UpdateExisting = false)
```
Comment (danmcc) [inline constructor.io/ConstructorIOAPI.cs]: The UpdateExisting flag is a good idea, but unfortunately none of our other libraries support it, and I think being consistent between them is important.  Sorry to cause you more work, but do you mind removing the UpdateExisting references?

File: ConstructorIO.Test/ListItemTests.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+        [TestMethod]
+        public void TestBatchAddBatchRemove()
+        {
+            var api = MakeAPI();
+
+            List<ListItem> batchTestSet = new List<ListItem>();
+
+            for (int i = 0; i < 5; i++)
+                batchTestSet.Add(new ListItem("Batch Add Batch Remove Test Item " + i, ListItemAutocompleteType.SearchSuggestions));
```
Comment (danmcc) [inline ConstructorIO.Test/ListItemTests.cs]: The *Batch API calls can only operate on one autocomplete_section at a time, so I don't think it makes sense to be able to add an autocomplete_section to a ListItem.

File: constructor.io/ConstructorIOAPI.cs (hunk 4 above)
Comment (ben-abraham) [inline constructor.io/ConstructorIOAPI.cs]: Yea, when I add the AddOrUpdate function, I will be removing the UpdateExisting flag.

File: ConstructorIO.Test/ListItemTests.cs (hunk 5 above)
Comment (ben-abraham) [inline ConstructorIO.Test/ListItemTests.cs]: I can add a constructor for ListItem that only takes one parameter for the Name. For batch items this doesn't cause a problem, but with the individual calls, there is potential for a server-side error.
Example:
`bool worked = api.Add(new ListItem("Test Item"))` this will error (server-side) and return false to the client because the new ListItem has no section.

It's not the end of the world, but something to be aware of.

File: constructor.io/ListItemAutocompleteType.cs (hunk 2 above)
Comment (danmcc) [inline constructor.io/ListItemAutocompleteType.cs]: Gotcha.  Sounds good -- thanks.

File: ConstructorIO.Test/ListItemTests.cs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+            Assert.IsTrue(api.Verify());
+        }
+
+        [TestMethod]
+        public void TestAddRemove()
+        {
+            var api = MakeAPI();
+
+            ListItem testItem = new ListItem("AddRemove Test Item", ListItemAutocompleteType.SearchSuggestions);
```
Comment (danmcc) [inline ConstructorIO.Test/ListItemTests.cs]: Could you please also add some tests where ListItems have metadata like url, description, and image_url?

File: ConstructorIO.Test/ListItemTests.cs (hunk 5 above)
Comment (danmcc) [inline ConstructorIO.Test/ListItemTests.cs]: The interfaces to Add() in the other libraries are (item, autocomplete_section).  How do you feel about that option?

Comment (ben-abraham) [inline ConstructorIO.Test/ListItemTests.cs]: I would really like to use the ListItem model across as many functions as possible for consistency, but I can add a helper function for Add (Remove too) that would take 2 strings and create/use the model for you. 
eg.
`var result = api.Add("Test", "Products")` internally just doing Add(new ListItem("Test", "Products"))

File: constructor.io/Logger.cs [hunk 7]
```diff
@@ -0,0 +1,114 @@
+﻿using System;
```
Comment (danmcc) [inline constructor.io/Logger.cs]: Could we call this file Tracker.cs?  I saw "logger" and thought it was a class for things like debug output.

Comment (ben-abraham) [inline constructor.io/Logger.cs]: Already renamed on my computer, just not pushed yet.

File: ConstructorIO.Test/ListItemTests.cs (hunk 5 above)
Comment (danmcc) [inline ConstructorIO.Test/ListItemTests.cs]: I see, that makes sense.  What about removing autocomplete_section from ListItem altogether, then?  That way, the section isn't really a part of the model -- it's more like you're saying "here's an item -- put it into this section" -- and it works across both single and batch methods.

Comment (danmcc) [inline ConstructorIO.Test/ListItemTests.cs]: Here's another wrinkle -- from our "Remove an Item" docs, where we explain the "id" parameter: "An arbitrary ID you optionally specified when adding the item. If passed in, you don’t need to pass in item_name."  So a ListItem doesn't necessarily need a name -- it can just be a privateId.  Can you think of a good way to create a ListItem interface that just takes a privateId and not a name?

**Comment by ben-abraham** [comment]
I'm still working on a few items, but I got the majority out of the way. Let me know if you have any questions.

**Comment by danmcc** [comment]
Awesome -- thanks!  Lookin' great so far.

**Comment by ben-abraham** [comment]
Added some more tests, and fixed some issues, so now all the tests are passing.

**Comment by danmcc** [comment]
Great, thanks!  Would you mind adding a quick test for the case where the name is null and only the PrivateID is passed in for removing items individually and in batches?  We have a client who is planning on using that functionality specifically.

**Comment by ben-abraham** [comment]
There, added 2 test cases for removing products by ID only both in bulk and sequentially.

**Comment by danmcc** [comment]
Great, thanks!  Does that wrap up everything?  If so, I'll do one last review and then push to nuget.

**Comment by ben-abraham** [comment]
I still have a few things I want to change/add yet, not too many though.
As far as the old example projects, what would you like done with them? I can update them to the new API, or we can consider the Test cases as replacement.

**Comment by ben-abraham** [comment]
Do you want me to implement the Query call as well?

**Comment by danmcc** [comment]
> I still have a few things I want to change/add yet, not too many though.

Cool.  No pressure, but any idea if you'll be able to complete them tonight?

> As far as the old example projects, what would you like done with them?

Since we have them, it'd be nice to update them if it's not too much trouble.

> Do you want me to implement the Query call as well?

No, that's ok.  Thanks, though.

File: constructor.io/ConstructorIOAPI.cs [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+
+        /// <summary>
+        /// Adds multiple items in one batch upload.
+        /// </summary>
+        /// <param name="Items">The items to add.</param>
+        /// <param name="AutocompleteSection">The section the items should be added to.</param>
+        /// <returns>true if successful</returns>
+        public bool AddBatch(IEnumerable<ListItem> Items, string AutocompleteSection)
+        {
+            return AddBatchAsync(Items, AutocompleteSection).GetAwaiter().GetResult();
```
Comment (adetch) [inline constructor.io/ConstructorIOAPI.cs]: Assuming this is a typo here?

Comment (ben-abraham) [inline constructor.io/ConstructorIOAPI.cs]: I guess I'm not seeing it, what part in specific looks like a typo?

**Comment by ben-abraham** [comment]
That's all of the changes I was looking to make. I did not have time yet to update the sample projects, which I will do tomorrow.

**Comment by danmcc** [comment]
Great, thanks!

**Comment by danmcc** [comment]
Hi Ben -- I was just going over ListItem() again.  Right now, if you call `new ListItem("string")`, it assumes "string" is a privateId, but `new ListItem("string", "section")` assumes "string" is a name, right?  That's a little confusing.  Any suggestions on how to make that more intuitive?

File: constructor.io/ConstructorIOAPI.cs (hunk 8 above)
Comment (adetch) [inline constructor.io/ConstructorIOAPI.cs]: I'm just a technical product manager who misread your code = )

**Comment by ben-abraham** [comment]
Yes, you are correct about that, the function header can be kind of confusing, however the documentation does make it fairly obvious. I'll send a picture in a little bit of what it looks like in Visual Studio, but I definitely see your point. If you think we need to change it, I see two possible ways,
1) `var item = new Item(PrivateID: "SomeID")`
2) `var item = Item.FromID("SomeID")`

Note: in case 1, there would be a single constructor with 3 optional params, so any of the following would be valid. Code is slightly more verbose, but very explicit/readable in what its doing.
`new Item(PrivateID: "SomeID")`
`new Item(Name: "Name", AutocompleteSection: "Section")`
`new Item(Name: "Name", AutocompleteSection: "Section", PrivateID: "SomeID")`

**Comment by danmcc** [comment]
Cool.  For case 1, could we expand that further to allow all parameters to be passed in?  e.g. `new Item(Name: "X", AutocompleteSection: "Y", Url: "http://...", ImageUrl: "http://...", etc)`

That would match our other libraries nicely.

**Comment by danmcc** [comment]
Awesome -- thank you!  Another thing we've been debating -- "PrivateID" is called "ID" in our other libraries and in our docs.  Could we change it to "ID" in this library?

**Comment by ben-abraham** [comment]
There, renamed the field.

**Comment by danmcc** [comment]
Awesome -- thank you!

**Comment by ben-abraham** [comment]
I still have not finished converting the two example projects, but we can leave them disabled and I can do it as another PR if you want to make these changes live now.

**Comment by danmcc** [comment]
Ok -- yeah, I'm going to merge these to master, and we can work from another PR from here.  Thanks!
